### DISCUSSION TITLE
How to subscribe to property changes and log the changes to console?

### DISCUSSION THREAD

**Discussion by DanOpcode** [AvaloniaUI/Avalonia#15965] [Q&A] [answered]
I'm reading the documentation on <https://docs.avaloniaui.net/docs/guides/data-binding/binding-from-code#subscribing-to-changes-to-a-property>. It gives this coding example:

```csharp
var textBlock = new TextBlock();
var text = textBlock.GetObservable(TextBlock.TextProperty);
text.Subscribe(value => Console.WriteLine(value + " Changed"));
```

However, when I try it, I get this error:

![image](https://github.com/AvaloniaUI/Avalonia/assets/980838/1b7f8cf4-f7df-41e9-ba8f-731760665bdf)

And when I try to make the lamba function's parameter explicitly of the type `string`, I get this error instead:

![image](https://github.com/AvaloniaUI/Avalonia/assets/980838/12568fac-1a92-48e2-93d7-6c494b9b1434)

Questions:

1. Is the documentation out-of-date?
2. How do I create an instance of IObserver that just logs the change to the console (if a lambda function is not possible to use anymore)?

**Comment by stevemonaco** [comment]
The docs are correct, but you do need `using Avalonia;` for the specialized `GetObservable` extension method.

You probably should be using `Debug.WriteLine` instead if you're looking for debug output. I don't see any output in VS with `Console.WriteLine` anyways.

**Comment by DanOpcode** [comment]
@stevemonaco Thanks! `GetObservable(..)` doesn't show any error though. My complete code (including the using statements) is:

```csharp
using Avalonia;
using Avalonia.Controls;
using System;

namespace AvaloniaApplication3;

public partial class MainWindow : Window
{
    public MainWindow()
    {
        InitializeComponent();
        var textBlock = new TextBlock();
        var text = textBlock.GetObservable(TextBlock.TextProperty);
        text.Subscribe(value => Console.WriteLine(value + " Changed"));
    }
}
```

The error is with:

```csharp
text.Subscribe(value => Console.WriteLine(value + " Changed"));
```

As you can see, there is a red squiggly line under `value`:

![image](https://github.com/AvaloniaUI/Avalonia/assets/980838/f077fe74-4dc0-4c4c-b688-afeabba9af6c)

**Reply by stevemonaco** [reply]
I can't replicate that. Clean solution, rebuild, and double-check that it's a real error (ie. if the project builds by looking at the output).  Your C# version is at least 10 (file-scoped namespaces), so the lambda handling shouldn't be that weak. You could try upgrading to C# 12 if you aren't already using it. 🤷‍♂️

**Comment by DanOpcode** [comment]
@stevemonaco I have uploaded my solution here: https://github.com/matachi/AvaloniaApplication3, so you can see my code. I created the project using one of the available templates, and it's .NET 8 so I assume it's C# 12 already. Can you see if I'm missing something obvious?

**Reply by stevemonaco** [reply, marked as answer]
Just tested. Needs System.Reactive package from NuGet and `using System.Reactive;`. The project I was testing in already had it added. You can either add the package directly or via ReactiveUI.

**Reply by DanOpcode** [reply]
Super! Installing `System.Reactive` made it work. Now it compiles!

**Reply by luizzeroxis** [reply]
If you don't want to install that, you can use `AnonymousObserver` from `Avalonia.Reactive`:

`text.Subscribe(new AnonymousObserver<string?>(value => Console.WriteLine(value + " Changed")));`

**Comment by ykarpeev** [comment]
I think you need the "Avalonia.ReactiveUI" package for that to work.

**Reply by DanOpcode** [reply]
Sweet, thanks! Seems like installing either the NuGet package `System.Reactive` or `Avalonia.ReactiveUI` does the job. Thank you very much!

**Comment by Koichi-hub** [comment]
I added `using System;` and it has worked. Maybe it will help someone
